File: app/src/main/java/com/ppy/nfcsample/adapter/BaseAdapter.java
package com.ppy.nfcsample.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;

/**
 * Created by ZP on 2017/8/8.
 */

public abstract class BaseAdapter<T> extends RecyclerView.Adapter<BaseViewHolder> {

    protected int mLayoutId;//布局id
    protected List<T> mDatas;//数据源
    protected Context mContext;//上下文
    private LayoutInflater mInflater;
    private MultipleTypeSupport<T> mMultipleTypeSupport;
    private OnItemClickListener mOnItemClickListener;
    private OnItemLongClickListener mOnItemLongClickListener;

    public BaseAdapter(List<T> datas, Context context, MultipleTypeSupport<T> multipleTypeSupport) {
        this(-1, datas, context);
        mMultipleTypeSupport = multipleTypeSupport;
    }

    public BaseAdapter(int layoutId, List<T> datas, Context context) {
        this.mContext = context;
        this.mLayoutId = layoutId;
        this.mDatas = datas;
        this.mInflater = LayoutInflater.from(mContext);
    }

    @Override
    public int getItemViewType(int position) {
        if (mMultipleTypeSupport != null) {
            return mMultipleTypeSupport.getLayoutId(mDatas.get(position), position);
        }
        return super.getItemViewType(position);
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (mMultipleTypeSupport != null) {
            mLayoutId = viewType;
        }
        View itemView = mInflater.inflate(mLayoutId, parent, false);
        return new BaseViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(final BaseViewHolder holder, final int position) {
        bindData(holder, mDatas.get(position), position);
        if (mOnItemClickListener != null) {
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    mOnItemClickListener.onItemClick(holder.itemView, position);
                }
            });
        }

        if (mOnItemLongClickListener != null) {
            holder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    mOnItemLongClickListener.onItemLongClick(holder.itemView, position);
                    return false;
                }
            });
        }
    }

    protected abstract void bindData(BaseViewHolder holder, T t, int position);

    @Override
    public int getItemCount() {
        return mDatas.size();
    }

    public interface OnItemClickListener {
        void onItemClick(View view, int position);
    }

    public interface OnItemLongClickListener {
        void onItemLongClick(View view, int position);
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        mOnItemClickListener = listener;
    }

    public void setOnItemLongClickListener(OnItemLongClickListener listener) {
        mOnItemLongClickListener = listener;
    }
}

File: app/src/main/java/com/ppy/nfcsample/adapter/BaseAdapter.kt
package com.ppy.nfcsample.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import androidx.recyclerview.widget.RecyclerView

/**
 * Created by ZP on 2017/8/8.
 */

abstract class BaseAdapter<T>(protected var mLayoutId: Int//布局id
                              , protected var mDatas: List<T>//数据源
                              , protected var mContext: Context//上下文
) : RecyclerView.Adapter<BaseViewHolder>() {
    private val mInflater: LayoutInflater
    private var mMultipleTypeSupport: MultipleTypeSupport<T>? = null
    private var mOnItemClickListener: OnItemClickListener? = null
    private var mOnItemLongClickListener: OnItemLongClickListener? = null

    constructor(datas: List<T>, context: Context, multipleTypeSupport: MultipleTypeSupport<T>) : this(-1, datas, context) {
        mMultipleTypeSupport = multipleTypeSupport
    }

    init {
        this.mInflater = LayoutInflater.from(mContext)
    }

    override fun getItemViewType(position: Int): Int {
        return mMultipleTypeSupport?.getLayoutId(mDatas[position], position)
                ?: super.getItemViewType(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
        if (mMultipleTypeSupport != null) {
            mLayoutId = viewType
        }
        val itemView = mInflater.inflate(mLayoutId, parent, false)
        return BaseViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: BaseViewHolder, position: Int) {
        bindData(holder, mDatas[position], position)
        mOnItemClickListener?.let {
            holder.itemView.setOnClickListener { mOnItemClickListener?.onItemClick(holder.itemView, position) }
        }
        if (mOnItemClickListener != null) {
            holder.itemView.setOnClickListener { mOnItemClickListener?.onItemClick(holder.itemView, position) }
        }

        if (mOnItemLongClickListener != null) {
            holder.itemView.setOnLongClickListener {
                mOnItemLongClickListener?.onItemLongClick(holder.itemView, position)
                false
            }
        }
    }

    protected abstract fun bindData(holder: BaseViewHolder, t: T, position: Int)

    override fun getItemCount(): Int {
        return mDatas.size
    }

    interface OnItemClickListener {
        fun onItemClick(view: View, position: Int)
    }

    interface OnItemLongClickListener {
        fun onItemLongClick(view: View, position: Int)
    }

    fun setOnItemClickListener(listener: OnItemClickListener) {
        mOnItemClickListener = listener
    }

    fun setOnItemLongClickListener(listener: OnItemLongClickListener) {
        mOnItemLongClickListener = listener
    }
}
